transform/utility_metric_freshness.go

```diff
@@ -0,0 +1,124 @@
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/brexhq/substation/config"
+	iconfig "github.com/brexhq/substation/internal/config"
+	"github.com/brexhq/substation/internal/errors"
+	"github.com/brexhq/substation/internal/metrics"
+	"github.com/brexhq/substation/message"
+)
+
+type utilityMetricFreshnessConfig struct {
+	Threshold string `json:"threshold"`
+
+	Object iconfig.Object `json:"object"`
+	Metric iconfig.Metric `json:"metric"`
+}
+
+func (c *utilityMetricFreshnessConfig) Decode(in interface{}) error {
+	return iconfig.Decode(in, c)
+}
+
+func (c *utilityMetricFreshnessConfig) Validate() error {
+	if c.Threshold == "" {
+		return fmt.Errorf("threshold: %v", errors.ErrMissingRequiredOption)
+	}
+
+	if c.Object.SourceKey == "" {
+		return fmt.Errorf("object_source_key: %v", errors.ErrMissingRequiredOption)
+	}
+
+	return nil
+}
+
+func newUtilityMetricFreshness(ctx context.Context, cfg config.Config) (*utilityMetricFreshness, error) {
+	conf := utilityMetricFreshnessConfig{}
+	if err := conf.Decode(cfg.Settings); err != nil {
+		return nil, fmt.Errorf("transform: utility_metric_freshness: %v", err)
+	}
+
+	if err := conf.Validate(); err != nil {
+		return nil, fmt.Errorf("transform: utility_metric_freshness: %v", err)
+	}
+
+	m, err := metrics.New(ctx, conf.Metric.Destination)
+	if err != nil {
+		return nil, fmt.Errorf("transform: utility_metric_freshness: %v", err)
+	}
+
+	dur, err := time.ParseDuration(conf.Threshold)
+	if err != nil {
+		return nil, fmt.Errorf("transform: utility_metric_freshness: duration: %v", err)
+	}
+
+	tf := utilityMetricFreshness{
+		conf:   conf,
+		metric: m,
+		dur:    dur,
+	}
+
+	return &tf, nil
+}
+
+type utilityMetricFreshness struct {
+	conf   utilityMetricFreshnessConfig
+	metric metrics.Generator
+	dur    time.Duration
+
+	success uint32
+	failure uint32
+}
+
+func (tf *utilityMetricFreshness) Transform(ctx context.Context, msg *message.Message) ([]*message.Message, error) {
+	// ctrl messages are handled by only one process, so the map
+	// updates below are safe for concurrency.
+	if msg.IsControl() {
+		tf.conf.Metric.Attributes["FreshnessType"] = "Success"
+		if err := tf.metric.Generate(ctx, metrics.Data{
+			Name:       tf.conf.Metric.Name,
+			Value:      tf.success,
+			Attributes: tf.conf.Metric.Attributes,
+		}); err != nil {
+			return nil, fmt.Errorf("transform: utility_metric_freshness: %v", err)
+		}
+
+		tf.conf.Metric.Attributes["FreshnessType"] = "Failure"
+		if err := tf.metric.Generate(ctx, metrics.Data{
+			Name:       tf.conf.Metric.Name,
+			Value:      tf.failure,
+			Attributes: tf.conf.Metric.Attributes,
+		}); err != nil {
+			return nil, fmt.Errorf("transform: utility_metric_freshness: %v", err)
+		}
+
+		atomic.StoreUint32(&tf.success, 0)
+		atomic.StoreUint32(&tf.failure, 0)
+		return []*message.Message{msg}, nil
+	}
+
+	// This is a time value expected to be in nanoseconds.
+	val := msg.GetValue(tf.conf.Object.SourceKey).Int()
```

Comment:
feat: Any reason to not support custom formatting of the time in the `SourceKey` object? If users are already transforming a timestamp field to ISO-8601 or another format this could reduce the amount of transforms and `meta` keys for measurement.

The main reason to not do this in my head was performance but if this is used in a configuration after send transforms then that small impact won't matter as much.

Reply:
I think the reason to not do is because it would be recreating [the existing `time.from` transforms](https://substation.readme.io/docs/tf-time). If you're already doing this:
```jsonnet

    sub.tf.time.from.str({ obj: { src: model.timestamp, trg: model.timestamp }, format: '2006-01-02T15:04:05Z' }),
    sub.tf.time.to.str({ obj: { src: model.timestamp, trg: model.timestamp }, format: const.time_fmt }),
```

Then I think this is a pretty trivial change:
```jsonnet

    sub.tf.time.from.str({ obj: { src: 'timestamp', trg: 'meta timestamp'}, format: '2006-01-02T15:04:05Z' }),
    sub.tf.time.to.str({ obj: { src: 'meta timestamp', trg: 'timestamp' }, format: const.time_fmt }),
```

Any datetime format conversion will (usually) "cost" two transforms (`from` and `to`), so by using the `meta` payload there's no additional work to do.

Reply:
That makes sense.  So if users change the target key location of the `time.from` then there are no extra transforms run. 👍

Reply:
Yeah, this requires zero (in the case of the above example) or one additional transforms.